### ISSUE TITLE
.Net 6. Running processes as separate tasks increases the execution time drastically

### ISSUE THREAD

**Issue by DanPristupov** [dotnet/runtime#67506]
### Description

Starting and running processes as separate tasks increases the execution time up to 10 times.

### Configuration

.Net 6.0
So far, I've reproduced the problem on both Intel and M1 macs.

The following code runs `ls`, but it can be any other process.

``` csharp
using System.Diagnostics;
using System.Text;

Console.WriteLine("Start processes sync");
Run(16, "/bin/ls", async: false);

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");


void Run(int count, string fileName, bool async)
{
    if (async)
    {
        var tasks = new Task[count];
        for (var i = 0; i < count; i += 1)
        {
            var task = new Task(() => {
                RunProcess(fileName);
            });
            tasks[i] = task;
            task.Start();
        }
        Task.WaitAll(tasks);
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

        process.WaitForExit();
    }
    Console.WriteLine($"{fileName}: {stopWatch.Elapsed.TotalMilliseconds}");
}
```

### Data

```
dotnet run --release
Start processes sync
/bin/ls: 103.4258
/bin/ls: 32.4446
/bin/ls: 31.3135
/bin/ls: 26.2724
/bin/ls: 32.4081
/bin/ls: 28.5405
/bin/ls: 51.8288
/bin/ls: 27.8595
/bin/ls: 26.0885
/bin/ls: 31.0516
/bin/ls: 29.5644
/bin/ls: 28.7161
/bin/ls: 28.6573
/bin/ls: 25.2447
/bin/ls: 27.5642
/bin/ls: 28.2451
Start processes async
/bin/ls: 179.849
/bin/ls: 179.9062
/bin/ls: 179.9018
/bin/ls: 179.9322
/bin/ls: 179.7889
/bin/ls: 179.818
/bin/ls: 179.9126
/bin/ls: 255.7262
/bin/ls: 255.3477
/bin/ls: 256.478
/bin/ls: 152.9311
/bin/ls: 152.9212
/bin/ls: 152.8825
/bin/ls: 152.9134
/bin/ls: 152.9793
/bin/ls: 153.5875
Done!
```

**Comment by ChrML** [comment]
What happens if you use threads instead?

I see there's a difference in the task vs non-task variant in that the async example starts all processes simultaneously (or as many as the threadpool will allow). While the sync variant starts them one by one sequentially. Meaning the problem could be anywhere, perhaps not related to tasks at all.

What happens if you modify the code to wait for each task one by one?

**Comment by danmoseley** [comment]
I ran this and consistently the child processes are slower -- even if I ensure the `ls` takes extra time by passing in `-R /home/`

How many cores do you have? I am guessing part of the issue is the the threadpool assumes you are giving it CPU bound work and will not allow more active threads than roughly the number of cores (or whatever its starting heuristic is). If launching and running `ls` is in fact mostly IO bound then this is a bad strategy.

Stepping back -- what are you trying to achieve?

**Comment by DanPristupov** [comment]
Thank you for your response.

@danmoseley 

> Stepping back -- what are you trying to achieve?

My use case is pretty simple. GUI application runs about 10-15 relatively short (~30-50ms) git commands on start which makes a noticeable freeze. The operations are (seemingly) absolutely independent. Surprisingly, after running them as tasks, the execution time of a single process increases multiple times and the total time remains basically the same.

While I was investigating the problem, I noticed, that the problem can be reproduced with any process, even with `/bin/ls` or `/bin/date`.

A few notes:

* My Intel MBP has 8 cores, the M1 Max machine has 10.

* I tried to increase the threadpool size, but that doesn't help. (btw, threadpool overflow usually looks different: some tasks execute normally and some tasks don't start. Instead we see that **all** processes are waiting for something).

* I also tried to run same operations first as warm up, but it makes no difference and I removed that code to make the example more simple.

* in debug mode, the slow down is even more noticeable.

@ChrML 

> I see there's a difference in the task vs non-task variant in that the async example starts all processes simultaneously (or as many as the threadpool will allow)
> Meaning the problem could be anywhere, perhaps not related to tasks at all.

It seems to me the problem is in the Process class, but I don't have enough expertise to confirm that (or not).

> What happens if you use threads instead?

I don't see any difference. Expand the following section for the details:
<details>
  <summary>Threads instead of Tasks. Increased Thread Pool</summary>
  
``` csharp
using System.Diagnostics;
using System.Text;

Console.WriteLine("Start processes sync");
Run(16, "/bin/ls", async: false);

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");


void Run(int count, string fileName, bool async)
{
    if (async)
    {
        ThreadPool.SetMinThreads(count + 10, count + 10);

        var threads = new Thread[count];
        for (var i = 0; i < count; i += 1)
        {
            threads[i] = new Thread(ThreadProc);
        }
        for (var i = 0; i < count; i += 1)
        {
            threads[i].Start(fileName);
        }
        for (var i = 0; i < count; i += 1)
        {
            threads[i].Join();
        }
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
}

void ThreadProc(object? obj)
{
    RunProcess((obj as string)!);
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

        process.WaitForExit();
    }
    Console.WriteLine($"{fileName}: {stopWatch.Elapsed.TotalMilliseconds}");
}
```

Results:
```
dotnet run --release
Start processes sync
/bin/ls: 101.7049
/bin/ls: 30.1693
/bin/ls: 27.4484
/bin/ls: 25.5091
/bin/ls: 28.6075
/bin/ls: 28.6546
/bin/ls: 28.2078
/bin/ls: 51.508
/bin/ls: 24.8652
/bin/ls: 28.4048
/bin/ls: 25.6099
/bin/ls: 24.9029
/bin/ls: 25.7239
/bin/ls: 26.31
/bin/ls: 25.509
/bin/ls: 24.9722
Start processes async
/bin/ls: 49.5739
/bin/ls: 138.7723
/bin/ls: 115.4152
/bin/ls: 140.6831
/bin/ls: 164.2454
/bin/ls: 165.0776
/bin/ls: 228.894
/bin/ls: 185.5192
/bin/ls: 300.4582
/bin/ls: 185.252
/bin/ls: 184.9457
/bin/ls: 185.1772
/bin/ls: 323.0178
/bin/ls: 208.1066
/bin/ls: 208.5017
/bin/ls: 209.9374
Done!

```
</details>



> What happens if you modify the code to wait for each task one by one?

This way the processes are executed as fast as they should. Please expand the section below for the details:

<details>
  <summary>Wait for each thread</summary>

``` csharp
using System.Diagnostics;
using System.Text;

Console.WriteLine("Start processes sync");
Run(16, "/bin/ls", async: false);

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");


void Run(int count, string fileName, bool async)
{
    if (async)
    {
        ThreadPool.SetMinThreads(count + 10, count + 10);

        var threads = new Thread[count];
        for (var i = 0; i < count; i += 1)
        {
            threads[i] = new Thread(ThreadProc);
        }
        for (var i = 0; i < count; i += 1)
        {
            threads[i].Start(fileName);
            threads[i].Join();
        }
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
}

void ThreadProc(object? obj)
{
    RunProcess((obj as string)!);
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

        process.WaitForExit();
    }
    Console.WriteLine($"{fileName}: {stopWatch.Elapsed.TotalMilliseconds}");
}

```

Results:
```
dotnet run --release
Start processes sync
/bin/ls: 92.785
/bin/ls: 21.1047
/bin/ls: 18.0883
/bin/ls: 19.9052
/bin/ls: 18.3877
/bin/ls: 20.165
/bin/ls: 18.944
/bin/ls: 19.4233
/bin/ls: 18.045
/bin/ls: 19.8717
/bin/ls: 37.9207
/bin/ls: 19.257
/bin/ls: 18.8873
/bin/ls: 21.1175
/bin/ls: 21.6771
/bin/ls: 19.7997
Start processes async
/bin/ls: 18.4013
/bin/ls: 19.3454
/bin/ls: 17.8117
/bin/ls: 19.2499
/bin/ls: 17.5375
/bin/ls: 18.1795
/bin/ls: 17.9971
/bin/ls: 19.0148
/bin/ls: 17.4245
/bin/ls: 19.1555
/bin/ls: 17.822
/bin/ls: 19.2319
/bin/ls: 18.1246
/bin/ls: 19.3281
/bin/ls: 19.161
/bin/ls: 19.4298
Done!


```
</details>


I think I've answered all your questions. Please let me know if I missed something.

**Comment by ChrML** [comment]
Nice details, and research!

Based on this, the slowdown should be related to the degree of parallelism and processes started simultaneously, and not tasks/threads by themselves. Since waiting for each task/thread seems to yield the same result. Maybe the application has to wait for a common lock or a blocking OS resource.

I don't have too much insight into the Process implementation under Linux. Maybe someone else have and can explain why this occurs. Could also be valuable to see if the same occurs on Windows.

**Comment by devsko** [comment]
@DanPristupov Can you please measure the total elapsed time for sync / async.

**Comment by devsko** [comment]
Windows:
```
Start processes sync
C:\Windows\System32\cmd.exe: 89,7152
C:\Windows\System32\cmd.exe: 49,1774
C:\Windows\System32\cmd.exe: 49,8015
C:\Windows\System32\cmd.exe: 50,6141
C:\Windows\System32\cmd.exe: 52,5395
C:\Windows\System32\cmd.exe: 50,8179
C:\Windows\System32\cmd.exe: 48,7166
C:\Windows\System32\cmd.exe: 50,4042
C:\Windows\System32\cmd.exe: 53,0987
C:\Windows\System32\cmd.exe: 57,6313
C:\Windows\System32\cmd.exe: 64,7421
C:\Windows\System32\cmd.exe: 52,3299
C:\Windows\System32\cmd.exe: 53,0856
C:\Windows\System32\cmd.exe: 52,6622
C:\Windows\System32\cmd.exe: 56,1951
C:\Windows\System32\cmd.exe: 52,1156
TOTAL sync: 913,2566
Start processes async
C:\Windows\System32\cmd.exe: 109,2501
C:\Windows\System32\cmd.exe: 115,1507
C:\Windows\System32\cmd.exe: 129,4584
C:\Windows\System32\cmd.exe: 133,1002
C:\Windows\System32\cmd.exe: 145,9811
C:\Windows\System32\cmd.exe: 155,326
C:\Windows\System32\cmd.exe: 163,7159
C:\Windows\System32\cmd.exe: 179,9286
C:\Windows\System32\cmd.exe: 149,8854
C:\Windows\System32\cmd.exe: 140,2502
C:\Windows\System32\cmd.exe: 169,6546
C:\Windows\System32\cmd.exe: 161,2332
C:\Windows\System32\cmd.exe: 129,9286
C:\Windows\System32\cmd.exe: 140,7953
C:\Windows\System32\cmd.exe: 152,6978
C:\Windows\System32\cmd.exe: 130,9094
TOTAL async: 319,4197
Done!
```

**Comment by danmoseley** [comment]
> I ran this and consistently the child processes are slower -- even if I ensure the ls takes extra time by passing in -R /home/

I just realized I used Linux. My M1 hasn't arrived yet. Does this repro on Linux for someone else? Or is somehow specific to Windows and Mac (?!)

What if you log the time (eg ticks) between when the loop starts, and when process.Start is called? presumably that will show a delay, which ideally in your case would be zero for each call.

**Comment by DanPristupov** [comment]
@danmoseley 

> What if you log the time (eg ticks) between when the loop starts, and when process.Start is called? presumably that will show a delay, which ideally in your case would be zero for each call.

<details>
  <summary>Source code</summary>

```csharp
using System.Diagnostics;
using System.Text;

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");

void Run(int count, string fileName, bool async)
{
    var stopWatch = Stopwatch.StartNew();
    if (async)
    {
        var tasks = new Task[count];
        for (var i = 0; i < count; i += 1)
        {
            var task = new Task(() => {
                Console.WriteLine($"delay: {stopWatch.ElapsedMilliseconds} ms");
                RunProcess(fileName);
            });
            tasks[i] = task;
            task.Start();
        }
        Task.WaitAll(tasks);
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
    var typeString = async ? "async" : "sync";
    Console.WriteLine($"Total type {typeString} {stopWatch.Elapsed.TotalMilliseconds}");
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

        process.WaitForExit();
    }
    Console.WriteLine($"{fileName}: {stopWatch.Elapsed.TotalMilliseconds}");
}

```
</details>

MBP16 M1 Max:
```
dotnet run --release
Start processes async
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
delay: 11 ms
/bin/ls: 412.8496
/bin/ls: 412.8208
/bin/ls: 412.8002
/bin/ls: 412.8127
/bin/ls: 412.8233
/bin/ls: 412.7961
/bin/ls: 412.819
/bin/ls: 412.8425
/bin/ls: 412.837
/bin/ls: 412.8167
delay: 431 ms
delay: 431 ms
delay: 431 ms
delay: 431 ms
delay: 432 ms
delay: 432 ms
/bin/ls: 200.0511
/bin/ls: 200.1575
/bin/ls: 200.0651
/bin/ls: 200.1079
/bin/ls: 200.1359
/bin/ls: 200.1537
Total type async 632.3171
Done!
```

My current machine has 10 cores, so the ThreadPool size is also 10. There are 16 tasks, and because of the ThreadPool overflow 2 virtual groups appeared: 10 and 6 simultaneous tasks.

We can see that each single process in a group somehow waits for the whole group of the parallel processes.

**Comment by fernandozago** [comment]
Not sure. But i ran it on windows 10 (Ryzen 3700x)
Command: CMD /c dir

```
Start processes sync
cmd: 40,7933
cmd: 26,7772
cmd: 25,3197
cmd: 23,502
cmd: 24,3611
cmd: 29,0904
cmd: 28,5455
cmd: 28,3005
cmd: 24,7782
cmd: 26,6514
cmd: 26,4133
cmd: 31,5373
cmd: 26,9445
cmd: 25,1255
cmd: 28,9803
cmd: 31,7677
Total type sync 458,4898
Start processes async
cmd: 37,7033
cmd: 40,5682
cmd: 37,5978
cmd: 49,1791
cmd: 51,8796
cmd: 53,8477
cmd: 58,7341
cmd: 60,0082
cmd: 63,2641
cmd: 49,1104
cmd: 56,1296
cmd: 55,8794
cmd: 68,7703
cmd: 58,1821
cmd: 57,7187
cmd: 59,6214
Total type async 102,5157
Done!
```

But i think it maybe something with multiple threads running at the same time.
If you change your code to run single instance at a time (like on the sync code), it returns nearly the same result for me.

```
Start processes sync
cmd: 44,3693
cmd: 26,5006
cmd: 23,815
cmd: 25,1439
cmd: 29,0849
cmd: 28,9684
cmd: 25,3439
cmd: 24,5733
cmd: 28,3871
cmd: 34,0787
cmd: 30,0783
cmd: 29,2867
cmd: 29,1504
cmd: 27,2343
cmd: 29,0069
cmd: 24,982
Total type sync 470,7642
Start processes async
cmd: 24,8412
cmd: 26,7439
cmd: 33,3553
cmd: 31,74
cmd: 28,5698
cmd: 27,5991
cmd: 31,2083
cmd: 28,8033
cmd: 26,6959
cmd: 23,7428
cmd: 22,9894
cmd: 23,6098
cmd: 23,5023
cmd: 23,6591
cmd: 22,7147
cmd: 22,8315
Total type async 425,0549
Done!
```

**Comment by DanPristupov** [comment]
I decided to run the profiler.

<details>
  <summary>Source code</summary>

```csharp
using System.Diagnostics;
using System.Text;

System.Threading.Thread.Sleep(5*1000); // sleep to allow to attach the profiler

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");

void Run(int count, string fileName, bool async)
{
    var stopWatch = Stopwatch.StartNew();
    if (async)
    {
        var tasks = new Task[count];
        for (var i = 0; i < count; i += 1)
        {
            var task = new Task(() => {
                RunProcess(fileName);
            });
            tasks[i] = task;
            task.Start();
        }
        Task.WaitAll(tasks);
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
    var typeString = async ? "async" : "sync";
    Console.WriteLine($"Total type {typeString} {stopWatch.Elapsed.TotalMilliseconds}");
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

        process.WaitForExit();
    }
    Console.WriteLine($"{fileName}: {stopWatch.Elapsed.TotalMilliseconds}");
}
```
</details>

Run the profiler on M1 Max:
```
% ~/.dotnet/tools/dotnet-trace collect -n StartProcessTes --format SpeedScope
No profile or providers specified, defaulting to trace profile 'cpu-sampling'

Provider Name                           Keywords            Level               Enabled By
Microsoft-DotNETCore-SampleProfiler     0x0000F00000000000  Informational(4)    --profile 
Microsoft-Windows-DotNETRuntime         0x00000014C14FCCBD  Informational(4)    --profile 

Process        : /Users/dan/src/StartProcessTest/bin/Debug/net6.0/StartProcessTest
Output File    : /Users/dan/src/StartProcessTest/StartProcessTest_20220408_102243.nettrace

[00:00:00:04]	Recording trace 460.164  (KB)
Press <Enter> or <Ctrl+C> to exit...

Trace completed.
Writing:	/Users/dan/src/StartProcessTest/StartProcessTest_20220408_102243.speedscope.json
Conversion complete
```

The hot point is the `Interop.Sys.ForkAndExecProcess` call in `runtime/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs`.

<img width="1118" alt="Screenshot 2022-04-08 at 10 39 09" src="https://user-images.githubusercontent.com/618115/162399804-91bf0abe-1e0f-4949-a3b4-a3b02ebb6d66.png">

Direct link to the source code: https://github.com/dotnet/runtime/blob/6a889d234267a4c96ed21d0e1660dce787d78a38/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L519

```csharp
                // Invoke the shim fork/execve routine.  It will create pipes for all requested
                // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
                // descriptors, and execve to execute the requested process.  The shim implementation
                // is used to fork/execve as executing managed code in a forked process is not safe (only
                // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
                int errno = Interop.Sys.ForkAndExecProcess(
                    resolvedFilename, argv, envp, cwd,
                    startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
                    setCredentials, userId, groupId, groups,
                    out childPid, out stdinFd, out stdoutFd, out stderrFd);
```

I've attached the full trace file here [StartProcessTest_20220408_102243.speedscope.json.zip](https://github.com/dotnet/runtime/files/8450157/StartProcessTest_20220408_102243.speedscope.json.zip)

You can unzip and open it yourself on https://speedscope.net

How can we go deeper into `Interop.Sys.ForkAndExecProcess`?

**Comment by DanPristupov** [comment]
There is a comment in ForkAndExecProcess which implies that allowing multiple processes to start concurrently was at least considered during initial the implementation.

https://github.com/dotnet/runtime/blob/6a889d234267a4c96ed21d0e1660dce787d78a38/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L502

```csharp
            // Lock to avoid races with OnSigChild
            // By using a ReaderWriterLock we allow multiple processes to start concurrently.
            s_processStartLock.EnterReadLock();
```

Maybe if has never been working properly or it could be a regression.

**Comment by GSPP** [comment]
Even if there were a global lock I'd expect the total time to be roughly the same. A global lock would at most reduce concurrency to 1, as before. Lock overhead can't be relevant here. Why is it 10x slower?

**Comment by tmds** [comment]
The reader lock gets taken while starting processes, the writer lock is taken when a process exits.

In the async case, different processes are starting simultaneous, so the `WaitForExit` is blocked by acquiring the writer lock.

For the sync case, you need to sum the elapsed times.
For the async case, they are measuring concurrent events.

If you move the `StopWatch` to the `Run` method you get this result:

```
$ dotnet run -c Release
Start processes sync
/bin/ls: 60.3433
Start processes async
/bin/ls: 18.8684
Done!
```

**Comment by tmds** [comment]
For some additional info, see https://github.com/dotnet/runtime/issues/25879.

**Comment by DanPristupov** [comment]
@tmds 
> In the async case, different processes are starting simultaneous, so the WaitForExit is blocked by acquiring the writer lock.

I'm might be missing something, but it looks like `WaitForExit` is not related. At least on the profiler graph in my previous post (https://github.com/dotnet/runtime/issues/67506#issuecomment-1092616906) it doesn't take a significant time (~20ms comparing to ~200ms of `ForkAndExecProcess`).

> If you move the `StopWatch` to the `Run` method you get this result:
> 
> ```
> $ dotnet run -c Release
> Start processes sync
> /bin/ls: 60.3433
> Start processes async
> /bin/ls: 18.8684
> Done!
> ```


Do you think like this?

<details>
  <summary>Source code</summary>

```csharp
using System.Diagnostics;
using System.Text;

Console.WriteLine("Start processes sync");
Run(16, "/bin/ls", async: false);

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");

void Run(int count, string fileName, bool async)
{
    var stopWatch = Stopwatch.StartNew();
    if (async)
    {
        var tasks = new Task[count];
        for (var i = 0; i < count; i += 1)
        {
            var task = new Task(() => {
                RunProcess(fileName);
            });
            tasks[i] = task;
            task.Start();
        }
        Task.WaitAll(tasks);
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
    var typeString = async ? "async" : "sync";
    Console.WriteLine($"Total time {typeString} {stopWatch.Elapsed.TotalMilliseconds}");
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

        process.WaitForExit();
    }
}
```
</details>

In my case, the total time doesn't change that much.

M1:
```
dotnet run -c Release
Start processes sync
Total time sync 444.3092
Start processes async
Total time async 383.883
Done!
```
</details>



> For the sync case, you need to sum the elapsed times. 
> For the async case, they are measuring concurrent events.

Well, yes, but I don't understand why each of the concurrent events take 5-10 times longer. It doesn't excuse the fact that user must wait for output of a single process 200ms instead of 30ms.

**Comment by tmds** [comment]
Sorry, I had overlooked this issue was for macOS specifically.

On macOS, process starts are serialized due to missing `pipe2`:

https://github.com/dotnet/runtime/blob/cde7799901885d47c6c9c12dc3e2166d1a33d90a/src/native/libs/System.Native/pal_process.c#L286-L297

**Comment by tmds** [comment]
> On macOS, process starts are serialized due to missing pipe2:

And for similar reasons, Windows has a lock also:

https://github.com/dotnet/runtime/blob/2087c07830770d3c7b8e8452e42a398b16ac0a0b/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L451-L455

**Comment by DanPristupov** [comment]
@tmds is there any possible workaround for that particular use case (i.e. running about 10 processes about 60ms each). I'm just trying to decrease the delay a user must wait for data to appear.

**Comment by DanPristupov** [comment]
> If you can remove the `WaitForExit` call, those Tasks won't end up waiting to acquire the writer lock which is being held by the processes that are starting sequentially. So the output of the first processes will be available sooner.

No. I removed it and the result is the same :(

With `WaitForExit` (Total time async `335ms`):
```
dotnet run -c Release
Start processes sync
/bin/ls: 98.0125
/bin/ls: 26.7105
/bin/ls: 20.9646
/bin/ls: 23.403
/bin/ls: 20.2838
/bin/ls: 20.8735
/bin/ls: 21.4462
/bin/ls: 24.2901
/bin/ls: 40.2885
/bin/ls: 20.1572
/bin/ls: 20.286
/bin/ls: 24.0907
/bin/ls: 20.8128
/bin/ls: 21.826
/bin/ls: 20.0096
/bin/ls: 24.9194
Total time sync 452.1408
Start processes async
/bin/ls: 158.3848
/bin/ls: 158.4169
/bin/ls: 158.3936
/bin/ls: 158.4039
/bin/ls: 158.3727
/bin/ls: 158.4184
/bin/ls: 158.4634
/bin/ls: 196.3038
/bin/ls: 196.3243
/bin/ls: 196.9643
/bin/ls: 126.9954
/bin/ls: 126.9804
/bin/ls: 127.0865
/bin/ls: 127.0429
/bin/ls: 127.0582
/bin/ls: 127.8567
Total time async 335.4064
Done!
```

Without `WaitForExit` (Total time async `400ms`):
```
dotnet run -c Release
Start processes sync
/bin/ls: 99.3795
/bin/ls: 28.0014
/bin/ls: 26.4949
/bin/ls: 23.8075
/bin/ls: 22.1625
/bin/ls: 28.9287
/bin/ls: 25.5122
/bin/ls: 42.618
/bin/ls: 25.3854
/bin/ls: 23.391
/bin/ls: 23.8038
/bin/ls: 23.0368
/bin/ls: 24.1766
/bin/ls: 23.3039
/bin/ls: 25.1493
/bin/ls: 23.2374
Total time sync 492.1862
Start processes async
/bin/ls: 139.389
/bin/ls: 139.3716
/bin/ls: 139.3997
/bin/ls: 139.409
/bin/ls: 139.4599
/bin/ls: 139.9483
/bin/ls: 213.2685
/bin/ls: 237.7443
/bin/ls: 237.7407
/bin/ls: 238.8761
/bin/ls: 46.5697
/bin/ls: 92.3975
/bin/ls: 121.6485
/bin/ls: 121.612
/bin/ls: 122.8951
/bin/ls: 151.5456
Total time async 400.4798
Done!
```


<details>
  <summary>Source code</summary>

```csharp
using System.Diagnostics;
using System.Text;

Console.WriteLine("Start processes sync");
Run(16, "/bin/ls", async: false);

Console.WriteLine("Start processes async");
Run(16, "/bin/ls", async: true);

Console.WriteLine("Done!");

void Run(int count, string fileName, bool async)
{
    var stopWatch = Stopwatch.StartNew();
    if (async)
    {
        var tasks = new Task[count];
        for (var i = 0; i < count; i += 1)
        {
            var task = new Task(() => {
                RunProcess(fileName);
            });
            tasks[i] = task;
            task.Start();
        }
        Task.WaitAll(tasks);
    }
    else
    {
        for (var i = 0; i < count; i += 1)
        {
            RunProcess(fileName);
        }
    }
    var typeString = async ? "async" : "sync";
    Console.WriteLine($"Total time {typeString} {stopWatch.Elapsed.TotalMilliseconds}");
}

void RunProcess(string fileName)
{
    var stopWatch = Stopwatch.StartNew();

    using (var process = new Process())
    {
        process.StartInfo = new ProcessStartInfo
        {
            FileName = fileName,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            ErrorDialog = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            StandardOutputEncoding = Encoding.UTF8,
        };

        process.Start();

        var output = process.StandardOutput.ReadToEnd();

//        process.WaitForExit();
    }
    Console.WriteLine($"{fileName}: {stopWatch.Elapsed.TotalMilliseconds}");
}
```
</details>

**Comment by tmds** [comment]
```
/bin/ls: 46.5697
/bin/ls: 92.3975
```
You get these lower values due to not waiting for exit.

**Comment by DanPristupov** [comment]
I was curious how that case is handled by **Swift**. It turned out on Swift async is always faster.

M1 (sync: `17ms`, async: `3ms`). Async is `5.65` times faster!
```
dan@MBP16-M1 StartProcessTest % swiftc program.swift 
dan@MBP16-M1 StartProcessTest % ./program
Start processes sync
/bin/ls 2ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 0ms
/bin/ls 1ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
/bin/ls 0ms
Total time sync 17ms
Start processes async
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 3ms
/bin/ls 1ms
/bin/ls 2ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 2ms
/bin/ls 3ms
Total time async 3ms
```

Intel (sync: `31ms`, async: `4ms`). Async is `7.75` times faster!
```
dan@MBP16-Intel StartProcessTest % swiftc program.swift
dan@MBP16-Intel StartProcessTest % ./program
Start processes sync
/bin/ls 2ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
/bin/ls 1ms
Total time sync 31ms
Start processes async
/bin/ls 2ms
/bin/ls 2ms
/bin/ls 3ms
/bin/ls 2ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
/bin/ls 3ms
Total time async 4ms
```

<details>
  <summary>Source code</summary>

```swift
import Foundation


run(count: 16, path: "/bin/ls", async: false)
run(count: 16, path: "/bin/ls", async: true)

func run(count: Int, path: String, async: Bool) {
    print("Start processes \(async ? "async" : "sync")")
    let start = DispatchTime.now()
    let operationQueue = OperationQueue()

    if !async {
        operationQueue.addOperation {
            for _ in 0..<count {
                runProcess(path: path)
            }
        }
    } else {
        for _ in 0..<count {
            operationQueue.addOperation {
                runProcess(path: path)
            }
        }
    }
    
    operationQueue.waitUntilAllOperationsAreFinished()

    let nanoTime = DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds
    let elapsed = Int(Double(nanoTime) / 1_000_000)

    print("Total time \(async ? "async" : "sync") \(elapsed)ms")
}

func runProcess(path: String) {
    let start = DispatchTime.now()
    let process = Process()
    process.launchPath = path
    let outputPipe = Pipe()
    process.standardOutput = outputPipe

    process.launch()

    let _ = outputPipe.fileHandleForReading.readDataToEndOfFile()

    process.waitUntilExit()

    let nanoTime = DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds
    let elapsed = Int(Double(nanoTime) / 1_000_000)
    print("\(path) \(elapsed)ms")
}
```
</details>

**Comment by tmds** [comment]
> I was curious how that case is handled by Swift. It turned out on Swift async is always faster.

This uses the macOS specific `POSIX_SPAWN_CLOEXEC_DEFAULT` which means child processes only inherit specified handles.

With .NET, any inheritable handle is available to any child.

There is an open issue for an API that would allow to specify the inheritance: https://github.com/dotnet/runtime/issues/13943.

However, because there may be a child `Process` that enable inheritance, the lock that causes the .NET process starts to be serialized on macOS must still be used (even for processes that disable inheritance).

To use `POSIX_SPAWN_CLOEXEC_DEFAULT` on macOS like Swift does, inheritance must be disabled process-wide for all children.

**Comment by adamsitnik** [comment]
In .NET 11 Preview 4 we have improved the performance of `Process.Start` on macOS arm64 by up to x100. Please refer to https://devblogs.microsoft.com/dotnet/process-api-improvements-in-dotnet-11/#improved-process-creation-on-apple-platforms for details.

> To use `POSIX_SPAWN_CLOEXEC_DEFAULT` on macOS like Swift does, inheritance must be disabled process-wide for all children.

We have also exposed the ability to use `POSIX_SPAWN_CLOEXEC_DEFAULT`, by using the https://devblogs.microsoft.com/dotnet/process-api-improvements-in-dotnet-11/#handle-inheritance API

Moreover, the `lock` is now not taken for the whole time of starting the process, but only for the time when the pipes are being created. And only when any redirection was requested:

https://github.com/dotnet/runtime/blob/7f58900c3cf86d52fdd70bc95d697ab50ab7475a/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L1194-L1235

Having said that, I am going to close the issue. If anybody can still observe major performance degradation when using .NET 11 Preview 4 on macOS, please leave a comment so I am going to re-open the issue (I am sorry, but I don't have an access to arm64 mac).
